acme/acme/crypto_util.py

```diff
@@ -161,31 +160,22 @@ def _pyopenssl_cert_or_req_san(cert_or_req):
     :rtype: `list` of `unicode`
 
     """
-    # constants based on implementation of
-    # OpenSSL.crypto.X509Error._subjectAltNameString
+    # constants based on PyOpenSSL certificate/CSR text dump
+    label = "DNS"
     parts_separator = ", "
     part_separator = ":"
-    extension_short_name = b"subjectAltName"
-
-    if hasattr(cert_or_req, 'get_extensions'):  # X509Req
-        extensions = cert_or_req.get_extensions()
-    else:  # X509
-        extensions = [cert_or_req.get_extension(i)
-                      for i in range(cert_or_req.get_extension_count())]
-
-    # pylint: disable=protected-access,no-member
-    label = OpenSSL.crypto.X509Extension._prefixes[OpenSSL.crypto._lib.GEN_DNS]
-    assert parts_separator not in label
     prefix = label + part_separator
+    title = "X509v3 Subject Alternative Name:"
 
-    san_extensions = [
-        ext._subjectAltNameString().split(parts_separator)
-        for ext in extensions if ext.get_short_name() == extension_short_name]
+    text = jose.ComparableX509(cert_or_req).dump(OpenSSL.crypto.FILETYPE_TEXT)
```

Comment:
Dumping to `FILETYPE_TEXT` might in the end be less hacky way of achieving the API of this function... However, I don't feel confident that this solution is good enough without testing it more carefully. I think we should add some test certs that would uncover problems, e.g.:
- very huge SAN list
- IDN